src/screentovnc.cpp

```diff
@@ -604,6 +613,181 @@ void ScreenToVnc::updateClientCursors(rfbScreenInfoPtr rfbScreen, bool emptyMous
     isEmptyMouse = emptyMouse;
 }
 
+void ScreenToVnc::uinputCreateKeyboardDevice()
+{
+  struct uinput_user_dev uud;
+
+  uinputKeyboardDeviceFD = open("/dev/uinput", O_WRONLY | O_NONBLOCK);
+
+  ioctl(uinputKeyboardDeviceFD, UI_SET_EVBIT, EV_KEY);
+  for (int i=0; i < 256; i++) {
+    ioctl(uinputKeyboardDeviceFD, UI_SET_KEYBIT, i);
+  }
+
+
+  memset(&uud, 0, sizeof(uud));
+  snprintf(uud.name, UINPUT_MAX_NAME_SIZE, "uinput old interface");
+  write(uinputKeyboardDeviceFD, &uud, sizeof(uud));
+
+  ioctl(uinputKeyboardDeviceFD, UI_DEV_CREATE);
+}
+
+void ScreenToVnc::emitKeystroke(int type, int code, int val)
+{
+   struct input_event ie;
+
+   ie.type = type;
+   ie.code = code;
+   ie.value = val;
+   /* timestamp values below are ignored */
+   ie.time.tv_sec = 0;
+   ie.time.tv_usec = 0;
+
+   write(uinputKeyboardDeviceFD, &ie, sizeof(ie));
+}
+
+void ScreenToVnc::keyboardHandler(rfbBool isKeyDown, rfbKeySym keySym, rfbClientPtr cl)
+{
+    int keyCode = -1;
+    int shift = 0;
```

Comment:
I'm wondering if this shift key handling should be done roughly like here or if it should do explicit shift presses. At least libvncserver library seems to have codes for shift keys, but I didn't check if clients use those.

Reply:
yea, clients i tested dont appear to send the shift key. would have made this code MUCH simpler.